### PR TITLE
Added basic sorting functionality

### PR DESCRIPTION
**Pull request by tacchinotacchi**
At the moment only logic is available, need tabs for graphic interactions

### CODE REVIEW

File: osu.Game/Screens/Select/CarouselContainer.cs [hunk 1]
```diff
@@ -184,6 +185,58 @@ public void SelectGroup(BeatmapGroup group, BeatmapPanel panel, bool animated =
             ScrollTo(selectedY, animated);
         }
 
+        public void Sort(FilterControl.SortMode mode) {
+            switch (mode) { 
+                case FilterControl.SortMode.Artist:
+                    groups.Sort((x, y) =>
+                    {
+                        return string.Compare(x.BeatmapSet.Metadata.Artist, y.BeatmapSet.Metadata.Artist);
```
Comment (Nabile-Rahmani) [inline osu.Game/Screens/Select/CarouselContainer.cs]: Can you replace these three lambda bodies by an expression ? (`groups.Sort((x, y) => string.Compare([...]);`)

File: osu.Game/Screens/Select/CarouselContainer.cs [hunk 2]
```diff
@@ -184,6 +185,58 @@ public void SelectGroup(BeatmapGroup group, BeatmapPanel panel, bool animated =
...
+                        return string.Compare(x.BeatmapSet.Metadata.Author, y.BeatmapSet.Metadata.Author);
+                    });
+                    break;
+                case FilterControl.SortMode.Difficulty:
+                    groups.Sort((x, y) =>
+                    {
+                        if (x.BeatmapSet.Beatmaps.First().BaseDifficulty.OverallDifficulty >
+                            y.BeatmapSet.Beatmaps.First().BaseDifficulty.OverallDifficulty)
+                            return 1;
+                        else if (Equals(x.BeatmapSet.Beatmaps.First().BaseDifficulty.OverallDifficulty,
```
Comment (Nabile-Rahmani) [inline osu.Game/Screens/Select/CarouselContainer.cs]: These two `else` can be dropped.

File: osu.Game/Screens/Select/CarouselContainer.cs [hunk 3]
```diff
@@ -184,6 +185,58 @@ public void SelectGroup(BeatmapGroup group, BeatmapPanel panel, bool animated =
...
+                case FilterControl.SortMode.Author:
+                    groups.Sort((x, y) =>
+                    {
+                        return string.Compare(x.BeatmapSet.Metadata.Author, y.BeatmapSet.Metadata.Author);
+                    });
+                    break;
+                case FilterControl.SortMode.Difficulty:
+                    groups.Sort((x, y) =>
+                    {
+                        if (x.BeatmapSet.Beatmaps.First().BaseDifficulty.OverallDifficulty >
```
Comment (Nabile-Rahmani) [inline osu.Game/Screens/Select/CarouselContainer.cs]: It's probably a good idea to leave a TODO comment there so we can replace `OverallDifficulty` by the actual star rating whenever it gets implemented.

**Review by Nabile-Rahmani — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Select/CarouselContainer.cs (hunk 1 above)
Comment (tacchinotacchi) [inline osu.Game/Screens/Select/CarouselContainer.cs]: I tried Compare, but for some reason it didn't have any effect. It's late in my timezone, so I'm gonna give it a look tomorrow

File: osu.Game/Screens/Select/CarouselContainer.cs (hunk 2 above)
Comment (tacchinotacchi) [inline osu.Game/Screens/Select/CarouselContainer.cs]: I was trying to follow the standard, but I'll drop them if the results are the same

File: osu.Game/Screens/Select/CarouselContainer.cs [hunk 4]
```diff
@@ -184,6 +185,49 @@ public void SelectGroup(BeatmapGroup group, BeatmapPanel panel, bool animated =
...
+                case FilterControl.SortMode.Difficulty:
+                    groups.Sort((x, y) =>
+                    {
+                        /*TODO: replace with star rating once implemented
+                        * Assumes BeatmapSets not to be grouped - or to be by difficulty,
+                        * otherwise this sorting makes little sense - or does it? */
+                        if (x.BeatmapSet.Beatmaps.First().BaseDifficulty.OverallDifficulty >
+                            y.BeatmapSet.Beatmaps.First().BaseDifficulty.OverallDifficulty)
+                            return 1;
+                        else
```
Comment (tacchinotacchi) [inline osu.Game/Screens/Select/CarouselContainer.cs]: IMO leaving one `else` improves readability. Feel free to change

File: osu.Desktop.VisualTests/Tests/TestCasePlaySongSelect.cs [hunk 5]
```diff
@@ -14,10 +15,17 @@ class TestCasePlaySongSelect : TestCase
     {
         private BeatmapDatabase db, oldDb;
         private TestStorage storage;
+        private Random rnd = new Random();
```
Comment (UselessToucan) [inline osu.Desktop.VisualTests/Tests/TestCasePlaySongSelect.cs]: [osu.Framework.MathUtils.RNG](https://github.com/ppy/osu-framework/blob/master/osu.Framework/MathUtils/RNG.cs)

**Review by UselessToucan — CHANGES_REQUESTED** [review]

File: osu.Desktop.VisualTests/Tests/TestCasePlaySongSelect.cs [hunk 6]
```diff
@@ -35,9 +43,19 @@ public override void Reset()
 
                 db.Import(sets);
             }
-            Add(new PlaySongSelect());
+
+            Add(songSelect = new PlaySongSelect());
+            OnArtist = () => { songSelect.Filter.Sort = FilterControl.SortMode.Artist; };
```
Comment (peppy) [inline osu.Desktop.VisualTests/Tests/TestCasePlaySongSelect.cs]: why even make these when you could put the anonymous delegates in the AddButton lines below?

File: osu.Game/Screens/Select/CarouselContainer.cs (hunk 4 above)
Comment (peppy) [inline osu.Game/Screens/Select/CarouselContainer.cs]: p.s. star ratings are implemented now.

File: osu.Game/Screens/Select/CarouselContainer.cs [hunk 7]
```diff
@@ -184,6 +185,49 @@ public void SelectGroup(BeatmapGroup group, BeatmapPanel panel, bool animated =
...
+                case FilterControl.SortMode.Author:
+                    groups.Sort((x, y) => string.Compare(x.BeatmapSet.Metadata.Author, y.BeatmapSet.Metadata.Author));
+                    break;
+                case FilterControl.SortMode.Difficulty:
+                    groups.Sort((x, y) =>
+                    {
+                        /*TODO: replace with star rating once implemented
+                        * Assumes BeatmapSets not to be grouped - or to be by difficulty,
+                        * otherwise this sorting makes little sense - or does it? */
+                        if (x.BeatmapSet.Beatmaps.First().StarDifficulty >
```
Comment (Nabile-Rahmani) [inline osu.Game/Screens/Select/CarouselContainer.cs]: You could compute the mean star difficulty value of the sets.

**Review by Nabile-Rahmani — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Select/CarouselContainer.cs [hunk 8]
```diff
@@ -191,6 +192,58 @@ public void SelectGroup(BeatmapGroup group, BeatmapPanel panel, bool animated =
...
+                case FilterControl.SortMode.Title:
+                    groups.Sort((x, y) => string.Compare(x.BeatmapSet.Metadata.Title, y.BeatmapSet.Metadata.Title));
+                    break;
+                case FilterControl.SortMode.Author:
+                    groups.Sort((x, y) => string.Compare(x.BeatmapSet.Metadata.Author, y.BeatmapSet.Metadata.Author));
+                    break;
+                case FilterControl.SortMode.Difficulty:
+                    groups.Sort((x, y) =>
+                    {
+                        float xAverage=0, yAverage=0;
```
Comment (Nabile-Rahmani) [inline osu.Game/Screens/Select/CarouselContainer.cs]: [Average](https://msdn.microsoft.com/en-us/library/system.linq.enumerable.average(v=vs.110).aspx) ?

**Review by Nabile-Rahmani — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Select/CarouselContainer.cs (hunk 8 above)
Comment (tacchinotacchi) [inline osu.Game/Screens/Select/CarouselContainer.cs]: As far as I know, Average, unlike Sort, does not accept delegates. If that's the case, Average would only work with integers, doubles, etc.

**Pull request merged**